test/events/users/user_deleted_event_test.rb

```diff
@@ -0,0 +1,30 @@
+# frozen_string_literal: true
+
+require 'test_helper'
+
+class Users::UserDeletedEventTest < ActiveSupport::TestCase
+  test 'the event is persisted with all the necessary attributes' do
+    account = FactoryBot.create(:simple_provider, provider_account: master_account)
+    user = FactoryBot.create(:user, account: account)
+
+    event = Users::UserDeletedEvent.create(user)
+    Rails.application.config.event_store.publish_event(event)
+
+    event_stored = EventStore::Repository.find_event!(event.event_id)
+    assert_equal user.id, event_stored.user_id
+    assert_equal event.metadata[:provider_id], master_account.id
+  end
+
+  test 'the user deleted event is persisted with the necessary attributes when its associations are already destroyed' do
+    provider = FactoryBot.create(:simple_provider)
+    account = FactoryBot.create(:simple_buyer, provider_account: provider)
+    user = FactoryBot.create(:user, account: account)
+
+    provider.delete
+
+    event = Users::UserDeletedEvent.create(user.reload)
+    Rails.application.config.event_store.publish_event(event)
+
+    assert_not_nil EventStore::Repository.find_event(event.event_id)
+  end
```

Comment:
A couple more thins we could test here:
1. when its a buyer user, assert that `provider_id` is the id of the provider
2. when the provider account no longer exists, assert that `tenant_id` of the user is used.

Reply:
Right, good catch 👍 Thanks 💪

Reply:
I don't see this testing for the fallback case of https://github.com/3scale/porta/blob/258efd67d682654fd527a655406185d8058728a4/app/events/users/user_deleted_event.rb#L10

Reply:
According to the implementation, if the user does not have a provider account (an anomaly only expected when a whole chain of objects is being deleted asynchronously and the provider account happened to get deleted before the buyer user object), then we should still have a `provider_id` in the metadata. In this case, `provider_id` comes from `user.tenant_id`.

If this is important, then it should be tested. That is what I meant in https://github.com/3scale/porta/pull/717/files/fd296ee0ecd5dde2f6d71dab4f575e44a9d57269#r268268568 (2nd test suggested).

```ruby
test "uses the user's tenant_id when the provider account no longer exists" do
  user = FactoryBot.create(:user)
  user.stubs(provider_account: nil, tenant_id: 'missing_provider_account.id')
  event = Users::UserDeletedEvent.create(user)
  assert_equal 'missing_provider_account.id', event.metadata[:provider_id]
end
```

Reply:
@guicassolato That's what this test called `'the user deleted event is persisted with the necessary attributes when its associations are already destroyed` is currently doing, no? 😕 Line 26 does `provider.delete` and line 28 does `user.reload` so it should pick the `provider_id` there from the `tenant_id`, and then the last line checks that it has the right value: `assert_equal provider_id, event.metadata.fetch(:provider_id)`

Reply:
You're right. It doesn't really test the specifics of the implementation (which is fetching the `tenant_id`) but I think I guess the important part is that metadata includes _a_ `provider_id` no matter what.

Yet, it does not test the case where the provider account _does_ exist then.

Maybe the test should be described _"metadata always includes a provider_id without or without the account having been destroyed already"_ (though I'd still rather see two separate tests for that.)